homeassistant/components/recorder/__init__.py

```diff
@@ -691,31 +806,28 @@ def _run_event_loop(self):
         # Use a session for the event read loop
         # with a commit every time the event time
         # has changed. This reduces the disk io.
-        while event := self.queue.get():
+        self.stop_requested = False
+        while not self.stop_requested:
+            task = self.queue.get()
             try:
-                self._process_one_event_or_recover(event)
+                self._process_one_task_or_recover(task)
             except Exception as err:  # pylint: disable=broad-except
-                _LOGGER.exception("Error while processing event %s: %s", event, err)
+                _LOGGER.exception("Error while processing event %s: %s", task, err)
 
         self._shutdown()
 
-    def _process_one_event_or_recover(self, event):
+    def _process_one_task_or_recover(self, task: RecorderTask):
         """Process an event, reconnect, or recover a malformed database."""
         try:
-            if self._process_one_task(event):
-                return
-            self._process_one_event(event)
-            return
+            return task.handler()
```

Comment:
Instead of holding a strong reference to the recorder instance in the task object, can we pass it here?
```suggestion
            return task.handler(self)
```

Reply:
That's indeed much better, thanks @bdraco 👍